### ISSUE TITLE
[prometheus] Make a mount which is defined in extraConfigmapMounts optional

### ISSUE THREAD

**Issue by tom932** [prometheus-community/helm-charts#100]
**Is your feature request related to a problem? Please describe.**

We want to mount multiple configmaps as alerting rules via `extraConfigmapMounts`, as described in this comment: https://github.com/helm/charts/issues/9254#issuecomment-443554083
However, one configmap `prometheus-default-alerts` is available by default (managed by the team that sets up prometheus, and is deployed before the helm chart), but another one, `prometheus-alerts`, should only be mounted if it is available (managed by another team). 

Currently, when I deploy prometheus, it cannot start, since the second configmap is not available. (`MountVolume.SetUp failed for volume "server-prometheus-alerts" : configmap "prometheus-alerts" not found`)

Since kubernetes seems to be able to mount configmaps only if they are available, at least stated in https://stackoverflow.com/a/48228418/2286108 , it would be nice if the helm chart would provide this functionality for mounts too.

**Describe the solution you'd like**
It would be great if there would be a flag like show in the code piece below, that would allow to mount configmaps only if they are available:

```
extraConfigmapMounts:
    - name: prometheus-default-alerts
      mountPath: /etc/config/alerts
      configMap: prometheus-default-alerts
      readOnly: true
    - name: prometheus-alerts
      mountPath: /etc/config/alerts
      configMap: prometheus-alerts
      readOnly: true
      optional: true                # <-- this one is new
```

If a similar functionality is already available, please let me know, but I didn't find a hint in values.yaml or in the documentation.

Additionally, this would or example also be a good idea for `alertmanager.configFromSecret` and other places where configmaps/secrets are mounted.

If additional information is required (e.g. current values.yaml), please let me know!

**Comment by monotek** [comment]
How should a container, that can't be started, check for a missing file?

**Comment by tom932** [comment]
@monotek sorry, maybe my request was not clear enough:

If there would be the requested functionality, the pod should start, but simply not mount the non-available configmap.

The not-starting is the current situation, which I want to get rid of.

**Comment by monotek** [comment]
So why do you not just don't define it, if it not exists?

**Comment by tom932** [comment]
@monotek because I don't want to roll out prometheus again when the other team decides to add its own alerting rules.

We do it like you suggest it currently (as a workaround, which is really unhandy):
  - We roll the chart out without the team-specific alerting rules, but only the default ones mounted.  <- work for my team
  - The team adds the alerting configmap. <- work for other team
  - We add the mount of the configmap in values.yml. <- work for my team
  - We roll out the helm chart again. <- work for my team
That is really unhandy. 

If my suggestion would be implemented, the workflow would be much easier:
  - We roll out the chart (with the optionally mounted configmap, which is not available, but no problem) <- work for my team
  - The team adds the alerting configmap <- work for other team
  - The team restarts prometheus, so the optional configmap gets mounted <- work for other team
So after the initial rollout, my team wouldn't have additional work after the configmap is added, which increases usablity of the chart.

So, as you can see, we found a workaround, but it feels complicated, is more work and requires more communication between teams, which could be optimized away if the chart would support something that is supported by kubernetes itself already.

**Comment by monotek** [comment]
Imho a feature to disable an already optional part would inflate the chart unnecessarily.
So i vote against such implementation.

**Comment by tom932** [comment]
@monotek I am sorry, but I am not sure if you got my idea correctly:

It should not disable an optional part, it should make the mount optional, so if the configmap is not available, it is simply not mounted. Nothing is disabled here.

It is just about using an existing k8s-feature in the chart :)

I think implementing this would improve the usability of the chart, and if the new flag "optional" defaults to false, the behavior would be like before. So it would be backward compatible. Additionally, if the flag is not used, it it would not inflate the chart or its usage.

Honestly, I don't see downsides of that feature (except the development effort of course ;) )

**Comment by djhoese** [comment]
I found this issue after looking for a similar option for these charts. What @tom932 is talking about is a relatively undocumented feature, but there is an issue discussing it here:

https://github.com/kubernetes/website/issues/9829

It sounds like if the volume is declared as optional that the mount path will be created but won't have any contents. What isn't clear to me is if you then create the ConfigMap later, does the Pod load that new ConfigMap and have access to its contents?

@monotek It sounds like this would be a one line change to the template that wouldn't change behavior for anyone except those wanting to define their ConfigMaps later.

**Comment by djhoese** [comment]
Actually, @tom932 it sounds like based on the video mentioned in this comment (from the thread you linked to), prometheus-operator should be able to pick up additional config maps based on labels: https://github.com/helm/charts/issues/9254#issuecomment-474450253

Video: https://www.linkedin.com/learning/kubernetes-monitoring-with-prometheus/creating-alerts-in-prometheus

**Comment by monotek** [comment]
Please create a pr. Default should be false...

**Comment by djhoese** [comment]
Due to my lack of understanding of prometheus operator I thought ConfigMaps were the best way for me to define my own set of rules. I have since learned that I can make PrometheusRule objects that (I think) will do what I need for now. It looks like there isn't the same kind of functionality for the alert manager, but I'll keep looking. It is a little confusing whether as someone who wants a reproducible configuration without affecting a pre-existing prometheus operator installation if I should create my own alert manager instance or not.

Bottom line: I shouldn't need this config map stuff for rules, but still trying to understand alert managers, sorry.
